File: python/Strings/day6/implementAtoi.py
INT_MAX = 2**31 - 1
INT_MIN = -2**31

def isAtoi(s: str) -> int:
    if s is None:
        return 0
    
    i = 0
    n = len(s)

    while i < n and s[i] == ' ':
        i += 1

    if i == n:
        return 0
    
    sign = 1
    if s[i] == '+' or s[i] == '-':
        if s[i] == '-':
            sign = -1
        i += 1

    ans = 0
    while i < n and s[i].isdigit():
        digit = ord(s[i]) - ord('0')
        ans = ans * 10 + digit

        if sign == 1 and ans > INT_MAX:
            return INT_MAX
        
        if sign == -1 and -ans < INT_MIN:
            return INT_MIN

        i += 1
    
    return sign * ans




# ---- main ----
if __name__ == "__main__":
    s = input("Enter a string: ")
    result = isAtoi(s)
    print("Converted Integer:", result)

File: java/Strings/Day6/ImplementAtoi.java
package Day6;
import java.util.*;

public class ImplementAtoi {
    public int isAtoi(String s) {
        if (s == null || s.length() == 0) return 0;

        int i = 0, n = s.length();

        while (i < n && s.charAt(i) == ' ') i++;

        if (i == n) return 0;

        int sign = +1;
        if (s.charAt(i) == '+' || s.charAt(i) == '-') {
            if (s.charAt(i) == '-') sign = -1;
            i++;
        }


        long ans = 0;
        int maxi = Integer.MAX_VALUE;
        int mini = Integer.MIN_VALUE;

        while (i < n && Character.isDigit(s.charAt(i))) {

            ans = ans * 10 + (s.charAt(i) - '0');

            if (sign == 1 && ans > maxi) return maxi;
            if (sign == -1 && -ans < mini) return mini;

            i++;
        }

        
        return (int)(ans * sign);
    }


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter a string: ");
        String s = sc.nextLine();

        ImplementAtoi obj = new ImplementAtoi();
        int result = obj.isAtoi(s);

        System.out.println("Converted Integer: " + result);

        sc.close();
    }
}